blockchain/blockindex.go

```diff
@@ -269,7 +269,8 @@ func newBlockIndex(db database.DB, chainParams *chaincfg.Params) *blockIndex {
 // This function is safe for concurrent access.
 func (bi *blockIndex) HaveBlock(hash *chainhash.Hash) bool {
 	bi.RLock()
-	_, hasBlock := bi.index[*hash]
+	node := bi.LookupNode(hash)
```

Comment:
`bi.LookupNode()` also tries to acquire the blockindex read lock. It'll work because it's a read lock but it ideally should be avoided. I think it's better to use `bi.index[*hash]`.